text/000-no-proofs-in-kernel.md

```diff
@@ -0,0 +1,206 @@
+
+  let join_safe_env l =
+    List.iter (function (_,(Opaque f,_)) -> ignore(Future.force f) | _ -> ()) l
+
+end
+```
+
+Remark: Before the STM things were very similar:
+- no `constr Future.t`, always `constr`
+- no `Future.chain f g`, just call `g f`
+- no need to have `join_safe_environment`, since it used to contain no `Future.t`
+
+The choice back then was to be very conservative in the kernel, and just
+sprinkle `lazy` here and there (`Future.t` was like lazy but with some wrapping
+to deal with imperative code).
+
+## New code
+
+```ocaml
+type name = string (* absolute name *)
+type constr = int
+type types = int
+let mkType = 0
+(* no Future.t *)
+
+module SafeTyping : sig
+
+  (* The opaque type of checked env *)
+  type safe_env
+  val empty_env : safe_env
+
+  (* The update functions *)
+  val add_constant :
+    name -> constr -> types -> safe_env -> safe_env
+  val add_opaque_constant :
+    name -> types -> safe_env -> safe_env
+  (* no need for add_axiom *)
+
+  (* Works for transparent proofs. For the opaque ones Print, Extraction & co look elsewhere *)
+  val get_constant_body : safe_env -> name -> constr
+
+  (* justifications for opaque_constants *)
+  type opaque_proof = Missing | Proof of constr
+
+  (* This time it should be probably called validate_safe_env *)
+  val join_safe_env : safe_env ->  opaque_proof list -> unit
+
+end = struct
+
+  type body = Transparent of constr | Opaque
+  type safe_env = (name * (body * types)) list
+  let empty_env = []
+
+  let typecheck : safe_env -> constr -> types -> unit  = assert false
+
+  let add_constant name t ty l =
+    assert(not(List.mem_assoc name l));
+    typecheck l ty mkType;
+    typecheck l t ty ;
+    (name, (Transparent t, ty)) :: l
+
+  let add_opaque_constant name ty l =
+    assert(not(List.mem_assoc name l));
+    typecheck l ty mkType;
+    (name, (Opaque, ty)) :: l
+
+  let get_constant_body l n =
+    match List.assoc n l with
+    | Transparent x, _ -> x
+    | Opaque, _ ->
+        failwith "opaque proofs are not in the kernel, ask Library of the STM"
+
+  type opaque_proof = Missing | Opaque of constr
+
+  let rec join_safe_env l p =
+    match l, p with
+    | (_,(Transparent _,_)) :: l, p -> join_safe_env l p
+    | (name,(Opaque,ty)) :: l, Missing :: p ->
+          prerr_endline (name ^ " is an axiom")
+    | (_,(Opaque,ty)) :: l, Proof t :: p ->
+          typecheck l t ty;
+          join_safe_env l p
+    | _ -> failwith "wrong opaque proofs list"
+
+end
+```
+
+Remark:
+- no `Future.t`, no `lazy`
+- the asynchronous computation of proofs is totally unknown to the kernel,
+  it is the upper layers, eg STM, that pass to it an opaque_proof list
+- one can join/validate a safe env many times, with more or less axioms
+  (today you can't easily change your mind from Qed and Admitted)
+
+# Alternatives
+
+Isabelle does what we do now (via promises), but does not let one
+repair a proof (you simply redo it from scratch). Actually it is
```

Comment:
What does "repair" mean?

Reply:
In CoqIDE you can repair a broken proof without invalidating the rest of the document.